Remove unnecessary deserialization in Replicated map

```
[...]
        at com.hazelcast.internal.serialization.impl.DataSerializer.read(DataSerializer.java:130)
        at com.hazelcast.internal.serialization.impl.DataSerializer.read(DataSerializer.java:47)
        at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:46)
        at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:170)
        at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:234)
        at com.hazelcast.replicatedmap.impl.record.DataReplicatedRecordStore.unmarshall(DataReplicatedRecordStore.java:38)
        at com.hazelcast.replicatedmap.impl.record.AbstractReplicatedRecordStore.put(AbstractReplicatedRecordStore.java:157)
        at com.hazelcast.replicatedmap.impl.operation.PutOperation.run(PutOperation.java:61)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:172)
        at com.hazelcast.spi.impl.operationexecutor.classic.ClassicOperationExecutor.runOnCallingThread(ClassicOperationExecutor.java:392)
        at com.hazelcast.spi.impl.operationexecutor.classic.ClassicOperationExecutor.runOnCallingThreadIfPossible(ClassicOperationExecutor.java:345)
        at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:259)
        at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:243)
        at com.hazelcast.spi.impl.operationservice.impl.Invocation.invokeInternal(Invocation.java:210)
        at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:180)
        at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:46)
        at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:65)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:118)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:98)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:127)
        at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPartitionSpecificRunnable(OperationThread.java:159)
        at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:142)
        at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:124)
        at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:99)
Caused by: java.lang.ClassNotFoundException: package.Class
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:137)
        at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:115)
        at com.hazelcast.nio.ClassLoaderUtil.newInstance(ClassLoaderUtil.java:68)
        at com.hazelcast.internal.serialization.impl.DataSerializer.read(DataSerializer.java:119)
        ... 23 more
```

Operations can work with Data, no reason to deserialize them. This makes it easier to use Replicated Map from a client.

--- a/hazelcast-client/src/test/java/com/hazelcast/client/replicatedmap/ClientReplicatedMapPutSerializationTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/replicatedmap/ClientReplicatedMapPutSerializationTest.java
@@ -1,0 +1,84 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.replicatedmap;
+
+import com.hazelcast.client.test.TestHazelcastFactory;
+import com.hazelcast.config.Config;
+import com.hazelcast.config.InMemoryFormat;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.ReplicatedMap;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.DataSerializable;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class ClientReplicatedMapPutSerializationTest extends HazelcastTestSupport {
+
+    TestHazelcastFactory factory = new TestHazelcastFactory();
+
+    @After
+    public void tearDown() throws Exception {
+        factory.terminateAll();
+    }
+
+    static AtomicInteger deSerializationCount = new AtomicInteger(0);
+
+    @Test
+    public void testPutShouldNotDeserializeData() throws Exception {
+        String mapName = randomName();
+        Config config = new Config();
+        config.getReplicatedMapConfig(mapName).setInMemoryFormat(InMemoryFormat.BINARY);
+        HazelcastInstance server = factory.newHazelcastInstance(config);
+        HazelcastInstance client = factory.newHazelcastClient();
+        ReplicatedMap<String, SerializationCountingData> map = client.getReplicatedMap(mapName);
+        String key = randomString();
+        SerializationCountingData value = new SerializationCountingData();
+        map.put(key, value);
+        map.put(key, value);
+
+        // only deserialized once in the proxy.
+        assertEquals(1, deSerializationCount.get());
+    }
+
+
+    static class SerializationCountingData implements DataSerializable {
+        public SerializationCountingData() {
+        }
+
+        @Override
+        public void writeData(ObjectDataOutput out) throws IOException {
+        }
+
+        @Override
+        public void readData(ObjectDataInput in) throws IOException {
+            deSerializationCount.incrementAndGet();
+        }
+    }
+}

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java
@@ -155,11 +155,10 @@ public abstract class AbstractReplicatedRecordStore<K, V> extends AbstractBaseRe
         } else {
             cancelTtlEntry(marshalledKey);
         }
-        Object unmarshalledOldValue = unmarshall(oldValue);
         if (replicatedMapConfig.isStatisticsEnabled()) {
             getStats().incrementPuts(Clock.currentTimeMillis() - time);
         }
-        return unmarshalledOldValue;
+        return oldValue;
     }
 
     @Override

--- a/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapPutSerializationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapPutSerializationTest.java
@@ -1,0 +1,78 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.replicatedmap;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.config.InMemoryFormat;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.ReplicatedMap;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.DataSerializable;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category(value = {QuickTest.class, ParallelTest.class})
+public class ReplicatedMapPutSerializationTest extends HazelcastTestSupport {
+
+    static AtomicInteger deSerializationCount = new AtomicInteger(0);
+
+    @Test
+    public void testPutShouldNotDeserializeData() throws Exception {
+        String mapName = randomName();
+        Config config = new Config();
+        config.getReplicatedMapConfig(mapName).setInMemoryFormat(InMemoryFormat.BINARY);
+        TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory();
+        HazelcastInstance instance1 = nodeFactory.newHazelcastInstance(config);
+        HazelcastInstance instance2 = nodeFactory.newHazelcastInstance(config);
+        ReplicatedMap<String, SerializationCountingData> map = instance1.getReplicatedMap(mapName);
+        String key = generateKeyOwnedBy(instance2);
+        SerializationCountingData value = new SerializationCountingData();
+        map.put(key, value);
+        map.put(key, value);
+
+        // only deserialized once in the proxy.
+        assertEquals(1, deSerializationCount.get());
+    }
+
+
+    static class SerializationCountingData implements DataSerializable {
+        public SerializationCountingData() {
+        }
+
+        @Override
+        public void writeData(ObjectDataOutput out) throws IOException {
+        }
+
+        @Override
+        public void readData(ObjectDataInput in) throws IOException {
+            deSerializationCount.incrementAndGet();
+        }
+    }
+
+}

